query-planner-js/src/buildPlan.ts

```diff
@@ -1256,65 +1153,52 @@ class FetchDependencyGraph {
+      this.mergeChildFetchesForSameSubgraphAndPath(group);
     }
 
-    this.isReduced = true;
+    this.mergeFetchesToSameSubgraphAndSameInputs();
+
+    this.isOptimized = true;
   }
 
   private removeEmptyGroups(group: FetchGroup) {
-    const dependents = this.dependents(group);
+    const children = group.children().concat();
     if (group.selection.isEmpty()) {
-      for (const dependent of dependents) {
-        // When we handle nested requires (@require of a field that has itself some @require), we may create temporary group
-        // for the "inner" require that never get any selection filled-in but are a dependency for the "outer" require. When,
-        // that happens however, the "outer" group should have dependency on other non-empty groups, and in that case it's
-        // safe to remove that one empty group (the group is useless and removing it doesn't break the dependency graph
-        // with remains connected).
-        assert(this.dependencies(dependent).length > 1, () => `Empty group ${group} is the *only* dependency of ${dependent}`);
-        this.removeDependency(dependent, group);
-      }
       this.remove(group);
     }
-    for (const g of dependents) {
+    for (const g of children) {
       this.removeEmptyGroups(g);
     }
   }
 
-  private mergeDependentFetchesForSameSubgraphAndPath(group: FetchGroup) {
-    const dependents = this.dependents(group);
-    if (dependents.length > 1) {
-      for (const g1 of dependents) {
-        for (const g2 of dependents) {
-          if (g1.index !== g2.index
-            && g1.subgraphName === g2.subgraphName
-            && sameMergeAt(g1.mergeAt, g2.mergeAt)
-            && this.dependencies(g1).length === 1
-            && this.dependencies(g2).length === 1
-          ) {
-            // We replace g1 by a new group that is the same except (possibly) for it's parentType ...
-            // (that's why we don't use `newKeyFetchGroup`, it assigns the index while we reuse g1's one here)
-            const merged = FetchGroup.create(this, g1.index, g1.subgraphName, g1.rootKind, g1.selection.parentType, g1.isEntityFetch, g1.mergeAt);
-            // Erase the pathsInParents as it's now potentially invalid (we won't really use it from that point on, but better
-            // safe than sorry).
-            this.pathsInParents[g1.index] = undefined;
-            if (g1.inputs) {
-              merged.addInputs(g1.inputs);
-            }
-            merged.addSelections(g1.selection);
-            this.groups[merged.index] = merged;
-
-            // ... and then merge g2 into that.
-            if (g2.inputs) {
-              merged.addInputs(g2.inputs);
-            }
-            merged.addSelections(g2.selection);
-            this.onMergedIn(merged, g2);
+  private mergeChildFetchesForSameSubgraphAndPath(group: FetchGroup) {
+    const children = group.children();
+    if (children.length > 1) {
+      for (const g1 of children) {
```

Comment:
This code makes me pretty nervous in that we seem to be iterating over an array that we're performing destructive operations on. Can we maybe move to a while loop with pointers?

Reply:
I hoped the `return` at the end of `if` would alleviate concerns, but the code was being inefficient anyway (restarting the iteration every time we merged, which is unnecessary), so switched to looping with indexes (not sure if that's what you meant by "while loop with pointers" but ...).